runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -5381,6 +5381,257 @@ bool genInstanceOfOrCheckCastNullTest(TR::Node* node, TR::CodeGenerator* cg, TR:
+      cg->getS390OutOfLineCodeSectionList().push_front(*outlinedSlowPath);
+      (*outlinedSlowPath)->swapInstructionListsWithCompilation();
+      }
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, callLabel);
+   // TODO : We can make this function a generic helper call function. Debug counter name then should be changed as per node.
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(comp, "instanceOf/(%s)/Helper", comp->signature()),1,TR::DebugCounter::Undetermined);
+   TR::S390CHelperLinkage *helperLink =  static_cast<TR::S390CHelperLinkage*>(cg->createLinkage(TR_CHelper));
+   return helperLink->buildDirectDispatch(node, deps, resultReg);
+   } 
+
+
+/** \brief
+ *     Generates a dynamicCache test with helper call for instanceOf/ifInstanceOf node
+ *
+ *  \details
+ *     This funcition generates a sequence to check per site cache for object class and cast class before calling out to jitInstanceOf helper
+ */
+static
+void genInstanceOfDynamicCacheAndHelperCall(TR::Node *node, TR::CodeGenerator *cg, TR::Register *castClassReg, TR::Register *objClassReg, TR::Register *resultReg, TR_S390ScratchRegisterManager *srm, TR::LabelSymbol *doneLabel, TR::LabelSymbol *helperCallLabel, TR::LabelSymbol *dynamicCacheTestLabel, TR::LabelSymbol *branchLabel, TR::LabelSymbol *trueLabel, TR::LabelSymbol *falseLabel, bool dynamicCastClass, bool generateDynamicCache, bool cacheCastClass, bool ifInstanceOf, bool trueFallThrough )
+   {
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = NULL;
+   TR::Compilation                *comp = cg->comp();
+   bool needResult = resultReg != NULL;
+   if (!castClassReg)
+      castClassReg = cg->evaluate(node->getSecondChild());
+   int32_t maxOnsiteCacheSlots = comp->getOptions()->getMaxOnsiteCacheSlotForInstanceOf();
+   TR::InstOpCode::Mnemonic loadHalfWordImmOp = TR::Compiler->target.is64Bit() ? TR::InstOpCode::LGHI : TR::InstOpCode::LHI;
+   bool needToGenerateDynamicCacheTest = generateDynamicCache && maxOnsiteCacheSlots > 0;
+   TR::Register *dynamicCacheReg = NULL;
+   int32_t addressSize = TR::Compiler->target.is64Bit() ? 8 : 4;
+   /* Layout of the writable data snippet
+    * Case - 1 : Cast class is runtime variable
+    * [UpdateIndex][ObjClassSlot-0][CastClassSlot-0]...[ObjClassSlot-N][CastClassSlot-N]
+    * Case - 2 : Cast Class is inetrface / unresolved
+    * [UpdateIndex][ObjClassSlot-0]...[ObjClassSlot-N]
+    * If there is only one cache slot, we will not have header.
+    * Last bit of cached objectClass will set to 1 indicating false cast
+    */
+   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * addressSize) + (addressSize * (maxOnsiteCacheSlots != 1));
+   if (needToGenerateDynamicCacheTest)
+      {
+      TR::S390WritableDataSnippet *dynamicCacheSnippet = NULL;
+      /* We can only request the snippet size of power 2, following table summarizes bytes needed for corresponding number of cache slots
```

Comment:
> We can only request the snippet size of power 2

Unrelated to this PR but this seems like an odd restriction. @joransiu do you might know of a historic reason why this is? Seems like we could benefit here by being able to allocate snippets on a word or double word boundary.